msg.go

```diff
@@ -1057,6 +1053,86 @@ func compressionLenSearch(c map[string]int, s string) (int, bool, int) {
 	return 0, false, fullSize + len(s)
 }
 
+// Truncate removes resource records from Additional, Ns, and Answer sections to
+// ensure the message doesn't exceed the given response buffer size. It doesn't
+// alter the message if it already fits. If resource records have to be removed
+// from the Answer section, the TC header bit will be set.
+//
+// Compression, EDNS0 and TSIG requirements are taken into account.
+func (dns *Msg) Truncate(size int) {
+	var lenFn = Len
+	if dns.Compress {
+		lenFn = newCompressionLenFn(dns.Question)
+	}
+
+	l := 12 // Message header size in bytes.
+	for _, q := range dns.Question {
+		l += q.len()
+	}
+
+	// OPT and TSIG records must be kept in the response.
+	for _, r := range dns.Extra {
+		switch r.Header().Rrtype {
+		case TypeOPT, TypeSIG:
+			l += lenFn(r)
+		}
+	}
+
+	var n int
+	n, l = maxRRs(dns.Answer, lenFn, l, size)
+	if n < len(dns.Answer) {
+		dns.Answer = dns.Answer[:n]
+		dns.Ns = nil
+		dns.Extra = filterExtra(dns.Extra, 0)
+		dns.Truncated = true
+		return
+	}
+
+	n, l = maxRRs(dns.Ns, lenFn, l, size)
+	if n < len(dns.Ns) {
+		dns.Ns = nil
+		dns.Extra = filterExtra(dns.Extra, 0)
+		return
+	}
+
+	n, l = maxRRs(dns.Extra, lenFn, l, size)
+	if n < len(dns.Extra) {
+		dns.Extra = filterExtra(dns.Extra, n)
+		return
+	}
+}
+
+// maxRRs returns the number of records fitting into the given message size
+// as well as the length of the resulting message including the given initial
+// size.
+func maxRRs(rs []RR, lenFn func(RR) int, initial, size int) (number, length int) {
+	i, l := 0, initial
+	for ; i < len(rs); i++ {
```

Comment:
this doesn't use binary search, so it's slower than the impl in coredns. Not sure if we need if-then to only do binary search of the number of RRs is larger than X

Reply:
The CoreDNS implementation does a binary search by calling Msg.Len() multiple times, so it calculates each answer/authority/additional resource record's len() multiple times. In this implementation the len() of each resource record is calculated at most once.

I don't understand how we could do a binary search here, we have to calculate the length of each record at least once as it's not fixed.